File: ortto_flutter_sdk_android/android/src/main/kotlin/com/ortto/messaging/flutter/OrttoFlutterSdkPlugin.kt
package com.ortto.messaging.flutter

import android.app.Activity
import android.app.Application
import android.content.Context
import android.util.Log
import com.google.firebase.messaging.RemoteMessage
import com.ortto.messaging.Ortto
import com.ortto.messaging.OrttoConfig
import com.ortto.messaging.PermissionUtil
import com.ortto.messaging.data.LinkUtm
import com.ortto.messaging.identity.UserID
import com.ortto.messaging.widget.CaptureConfig
import com.ortto.messaging.PushNotificationHandler

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

enum class PushPermission {
    ASK,
    DISABLED,
    GRANTED,
    PREVIOUSLY_DENIED,
    PREVIOUSLY_GRANTED,
}

class OrttoFlutterSdkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private var tag = "OrttoFlutterSdkPlugin";
    private lateinit var channel: MethodChannel
    private var applicationContext: Context? = null
    private var activity: Activity? = null;

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "ortto_flutter_sdk_android")
        channel.setMethodCallHandler(this)
        applicationContext = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getPlatformName" -> result.success(getPlatformName())
            "initialize" -> result.success(initialize(call))
            "initializeCapture" -> result.success(initializeCapture(call))
            "identify" -> result.success(identify(call))
            "requestPermissions" -> requestPermissions(result)
            "onMessageReceived" -> onMessageReceived(call)
            "queueWidget" -> {
                queueWidget(call)
                result.success(null)
            }
            "showWidget" -> {
                showWidget(call)
                result.success(null)
            }
            "processNextWidgetFromQueue" -> {
                processNextWidgetFromQueue()
                result.success(null)
            }
            "dispatchPushRequest" -> {
                dispatchPushRequest()
                result.success(null)
            }
            "clearData" -> {
                clearData()
                result.success(null)
            }
            "clearIdentity" -> clearIdentity(result)
            "trackLinkClick" -> trackLinkClick(call, result)
            else -> result.notImplemented()
        }
    }

    // This is the method you want to implement
    private fun getPlatformName(): String {
        return "Android"
    }

    private fun initialize(call: MethodCall): Unit? {
        val config = OrttoConfig(
            call.argument("appKey"),
            call.argument("endpoint"),
            call.argument("shouldSkipNonExistingContacts"),
            call.argument("allowAnonUsers"),
        )

        Ortto.instance().init(config, this.applicationContext as Application)

        return null;
    }

    private fun initializeCapture(call: MethodCall): Unit? {
        val config = CaptureConfig(
            call.argument("dataSourceKey"),
            call.argument("captureJsUrl"),
            call.argument("apiHost"),
        )

        Ortto.instance().initCapture(config)

        if (activity != null) {
            Ortto.instance().capture.setActivity(activity);
        }

        return null;
    }

    private fun identify(call: MethodCall): Unit? {
        val user = UserID.make();
        user.firstName = call.argument("firstName");
        user.lastName = call.argument("lastName");
        user.email = call.argument("email");
        user.acceptsGdpr = call.argument<Boolean>("acceptsGdpr") ?: false;
        user.contactId = call.argument("contactId");
        user.phone = call.argument("phone");
        user.externalId = call.argument("externalId");

        Ortto.instance().identify(user);

        return null;
    }

    private fun clearData() {
        Ortto.instance().clearData()
    }

    private fun dispatchPushRequest() {
        Ortto.instance().dispatchPushRequest()
    }

    private fun requestPermissions(result: MethodChannel.Result) {
        Ortto.instance().requestPermissions(activity, object : PermissionUtil.PermissionAskListener {
            override fun onPermissionAsk() {
                result.success(PushPermission.ASK.name)
            }

            override fun onPermissionDisabled() {
                result.success(PushPermission.DISABLED.name)
            }

            override fun onPermissionGranted() {
                result.success(PushPermission.GRANTED.name)
            }

            override fun onPermissionPreviouslyDenied() {
                result.success(PushPermission.PREVIOUSLY_DENIED.name)
            }

            override fun onPermissionPreviouslyGranted() {
                result.success(PushPermission.PREVIOUSLY_GRANTED.name)
            }
        })
    }

    private fun trackLinkClick(call: MethodCall, result: MethodChannel.Result) {
        val link = call.argument<String>("link");

        Ortto.instance().trackLinkClick(link
        ) {
            Log.d(tag, "trackLinkClick: $it")

            val map = mapOf(
                "utm_campaign" to it.campaign,
                "utm_medium" to it.medium,
                "utm_source" to it.source,
                "utm_content" to it.content
            )

            result.success(map);
        }
    }

    private fun queueWidget(call: MethodCall) {
        val widgetId = call.argument<String>("widgetId");

        Ortto.instance().queueWidget(widgetId);
    }

    private fun showWidget(call: MethodCall) {
        val widgetId = call.argument<String>("widgetId");

        Ortto.instance().showWidget(widgetId);
    }

    private fun processNextWidgetFromQueue() {
        Ortto.instance().capture?.processNextWidgetFromQueue()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    private fun onMessageReceived(call: MethodCall): Boolean {
        val message = call.argument<Map<String, Any>?>("message")
        val context = applicationContext as Context

        // Inline transformation logic
        val notification = message?.get("notification") as? Map<String, Any>
        val data = message?.get("data") as? Map<String, Any>
        val messageParams = mutableMapOf<String, Any>()

        notification?.forEach { (key, value) ->
            messageParams[key] = value
        }
        // Ensuring 'data' params take precedence as they are mainly used in rich push
        data?.forEach { (key, value) ->
            messageParams[key] = value
        }

        // Construct RemoteMessage
        val remoteMessageBuilder = RemoteMessage.Builder("test").apply {
            messageParams.forEach { (key, value) ->
                // if value is null, dont add it
                if (value != null) {
                    addData(key, value.toString())
                }
            }
            (message?.get("messageId") as? String)?.let { messageId = it }
            (message?.get("messageType") as? String)?.let { messageType = it }
            (message?.get("collapseKey") as? String)?.let { collapseKey = it }
            (message?.get("ttl") as? Int)?.let { ttl = it }
        }
        val remoteMessage = remoteMessageBuilder.build()

        val handler = PushNotificationHandler(remoteMessage)

        return handler.handleMessage(context)
    }

    private fun clearIdentity(result: MethodChannel.Result) {
        Ortto.instance().clearIdentity() {
            val responseMap = mapOf("sessionId" to it.sessionId)
            result.success(responseMap)
        }
    }
}

File: ortto_flutter_sdk_ios/ios/Classes/OrttoFlutterSdkPlugin.swift
import Flutter
import UIKit
import OrttoSDKCore
import OrttoPushMessagingFCM
import OrttoInAppNotifications
import OrttoPushMessaging
import FirebaseMessaging

public class OrttoFlutterSdkPlugin: NSObject, FlutterPlugin, UNUserNotificationCenterDelegate {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "ortto_flutter_sdk_ios", binaryMessenger: registrar.messenger())
        let instance = OrttoFlutterSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformName":
            result("iOS")
        case "initialize":
            initialize(call, result)
        case "initializeCapture":
            initializeCapture(call, result)
        case "identify":
            identify(call, result)
        case "clearData":
            clearData()
            result(nil)
        case "dispatchPushRequest":
            dispatchPushRequest()
            result(nil)
        case "requestPermissions":
            requestPermissions(result)
        case "registerDeviceToken":
            registerDeviceToken(call, result)
        case "trackLinkClick":
            trackLinkClick(call, result)
        case "queueWidget":
            queueWidget(call, result)
        case "showWidget":
            showWidget(call, result)
        case "processNextWidgetFromQueue":
            processNextWidgetFromQueue()
            result(nil)
        case "onMessageReceived":
            onMessageReceived(call, result)
        case "clearIdentity":
            clearIdentity(call, result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let configMap = call.arguments as? [String: Any?],
              let appKey = configMap["appKey"] as? String,
              !appKey.isEmpty else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "initialize requires a non-empty appKey",
                details: nil
            ))
            return
        }

        Ortto.initialize(
            appKey: appKey,
            endpoint: configMap["endpoint"] as? String,
            shouldSkipNonExistingContacts: configMap["shouldSkipNonExistingContacts"] as? Bool ?? false
        )
        result(nil)
    }

    private func initializeCapture(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let configMap = call.arguments as? [String: Any?],
              let dataSourceKey = configMap["dataSourceKey"] as? String,
              !dataSourceKey.isEmpty else {
            result(invalidArguments("initializeCapture requires a non-empty dataSourceKey"))
            return
        }

        do {
            try OrttoCapture.initialize(
                dataSourceKey: dataSourceKey,
                captureJsURL: url(from: configMap["captureJsUrl"]),
                apiHost: url(from: configMap["apiHost"])
            )
            result(nil)
        } catch {
            result(FlutterError(
                code: "CAPTURE_INITIALIZATION_ERROR",
                message: error.localizedDescription,
                details: String(describing: error)
            ))
        }
    }

    private func identify(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let userData = call.arguments as? [String: Any?] else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "identify requires an identity map",
                details: nil
            ))
            return
        }

        var user = UserIdentifier(
            contactID: userData["contact_id"] as? String,
            email: userData["email"] as? String,
            phone: userData["phone"] as? String,
            externalID: userData["external_id"] as? String,
            firstName: userData["first_name"] as? String,
            lastName: userData["last_name"] as? String
        )
        user.acceptsGDPR = userData["accepts_gdpr"] as? Bool ?? false

        Ortto.shared.identify(user) { response in
            switch response {
            case .success:
                result(nil)
            case .failure(let error):
                result(FlutterError(
                    code: "IDENTIFY_ERROR",
                    message: error.localizedDescription,
                    details: String(describing: error)
                ))
            }
        }
    }

    private func queueWidget(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any?],
              let widgetId = args["widgetId"] as? String,
              !widgetId.isEmpty else {
            result(invalidArguments("queueWidget requires a non-empty widgetId"))
            return
        }

        OrttoCapture.shared.queueWidget(widgetId)
        result(nil)
    }

    private func registerDeviceToken(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any?],
              let token = args["token"] as? String,
              !token.isEmpty else {
            result(invalidArguments("registerDeviceToken requires a non-empty token"))
            return
        }

        PushMessaging.shared.registerDeviceToken(fcmToken: token)
        result(nil)
    }

    private func trackLinkClick(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any?] else {
            result(invalidArguments("trackLinkClick requires an argument map"))
            return
        }

        guard let link = args["link"] as? String, !link.isEmpty else {
            result(invalidArguments("trackLinkClick requires a non-empty link"))
            return
        }

        guard let components = URLComponents(string: link),
              components.scheme != nil else {
            result(FlutterError(code: "INVALID_LINK", message: "The link is malformed", details: link))
            return
        }

        let trackingValue = components.queryItems?.first(where: { $0.name == "tracking_url" })?.value
        let utmSource = trackingValue.flatMap(decodeBase64URL) ?? link

        let linkUtm = Ortto.shared.retrieveUtmParameters(utmSource) ?? LinkUtm([])

        let response = linkUtmMap(linkUtm)
        guard trackingValue != nil else {
            result(response)
            return
        }

        var completed = false
        let complete: (Any) -> Void = { value in
            DispatchQueue.main.async {
                guard !completed else { return }
                completed = true
                result(value)
            }
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + 30) {
            guard !completed else { return }
            completed = true
            result(FlutterError(
                code: "TRACKING_ERROR",
                message: "Link tracking did not complete",
                details: link
            ))
        }

        Ortto.shared.trackLinkClick(link) {
            complete(response)
        }
    }

    private func decodeBase64URL(_ value: String) -> String? {
        var base64 = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder != 0 {
            base64.append(String(repeating: "=", count: 4 - remainder))
        }

        guard let data = Data(base64Encoded: base64) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func linkUtmMap(_ utm: OrttoSDKCore.LinkUtm) -> [String: Any] {
        [
            "utm_campaign": flutterValue(utm.campaign),
            "utm_medium": flutterValue(utm.medium),
            "utm_source": flutterValue(utm.source),
            "utm_content": flutterValue(utm.content)
        ]
    }

    private func flutterValue(_ value: String?) -> Any {
        if let value = value {
            return value
        }
        return NSNull()
    }

    private func requestPermissions(_ result: @escaping FlutterResult) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            DispatchQueue.main.async {
                switch settings.authorizationStatus {
                case .notDetermined:
                    result("ASK")
                case .denied:
                    result("PREVIOUSLY_DENIED")
                case .authorized, .provisional, .ephemeral:
                    result("PREVIOUSLY_GRANTED")
                @unknown default:
                    result(FlutterError(
                        code: "UNKNOWN_PERMISSION_STATUS",
                        message: "iOS returned an unknown notification authorization status",
                        details: settings.authorizationStatus.rawValue
                    ))
                }
            }
        }
    }

    private func invalidArguments(_ message: String) -> FlutterError {
        FlutterError(code: "INVALID_ARGUMENTS", message: message, details: nil)
    }

    private func url(from value: Any?) -> URL? {
        guard let string = value as? String, !string.isEmpty else {
            return nil
        }
        return URL(string: string)
    }

    private func onMessageReceived(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        result(false)
    }

    private func processNextWidgetFromQueue() {
        OrttoCapture.shared.processNextWidgetFromQueue()
    }

    private func dispatchPushRequest() {
        Ortto.shared.dispatchPushRequest()
    }

    private func clearData() {
        Ortto.shared.clearData()
    }

    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        let handled = PushMessaging.shared.userNotificationCenter(center, didReceive: response, withCompletionHandler: completionHandler)

        if !handled {
            completionHandler()
        }
    }

    private func showWidget(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let widgetId = args["widgetId"] as? String else {
            result([
                "success": false,
                "message": "Invalid or missing widgetId"
            ])
            return
        }

        guard let capture = OrttoCapture.shared else {
            result([
                "success": false,
                "message": "OrttoCapture not initialized"
            ])
            return
        }

        capture.showWidget(widgetId).then { showResult in
            switch showResult {
            case .success:
                result([
                    "success": true,
                    "message": "Widget shown successfully"
                ])
            case .failure(let error):
                if let widgetError = error as? WidgetError {
                    result([
                        "success": false,
                        "message": widgetError.errorDescription ?? "Widget error"
                    ])
                } else {
                    result([
                        "success": false,
                        "message": error.localizedDescription
                    ])
                }
            }
        }
    }

    public func clearIdentity(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        Ortto.shared.clearIdentity { response in
            if let response = response {
                let responseDict: [String: Any] = [
                    "sessionId": response.sessionId,
                    "success": true
                ]
                result(responseDict)
            } else {
                result([
                    "success": false,
                    "message": "No identity found or failed to unregister"
                ])
            }
        }
    }
}
